compiler/wasm/src/compile.rs

```diff
@@ -9,29 +9,54 @@ use noirc_driver::{
     add_dep, compile_contract, compile_main, prepare_crate, prepare_dependency, CompileOptions,
     CompiledContract, CompiledProgram,
 };
-use noirc_frontend::{graph::CrateGraph, hir::Context};
-use std::path::Path;
+use noirc_frontend::{
+    graph::{CrateGraph, CrateId, CrateName},
+    hir::Context,
+};
+use serde::Deserialize;
+use std::{collections::HashMap, path::Path};
 use wasm_bindgen::prelude::*;
 
-use crate::errors::JsCompileError;
+use crate::errors::{CompileError, JsCompileError};
 
 const BACKEND_IDENTIFIER: &str = "acvm-backend-barretenberg";
 
+#[wasm_bindgen(typescript_custom_section)]
+const DEPENDENCY_GRAPH: &'static str = r#"
+export type DependencyGraph = {
+    root_dependencies: readonly string[];
+    library_dependencies: Readonly<Record<string, readonly string[]>>;
```

Comment:
I'm not gonna ask for this to be changed so late in thie PR but we probably should have just gone for a proper tree here where we can properly represent separate dependency trees of libraries.

Just a heads up that we're going to have to address this in a separate breaking change to address this in future.

Reply:
I actually did come up with _the perfect_ typescript type for this but decided against it because:
(1) I didn't know how to tell wasm-bindgen to give `compile` type parameters
(2) the intended use case is for the dependency graph to be computed at runtime (from a bunch `Nargo.toml` files) in which case the compiler won't be able to enforce the type

For reference this was the type. It only allows known library names (dependencies of the root or of other libraries) and disallows dependencies on self (ie. foo can't depend on foo). The type permits cyclical dependencies though and duplicate entries (although the compiler handles this gracefully). Intellisense works well with this

```
type DependencyGraph<T extends string = string, E extends T | string = string> = {
  entrypoint: ReadonlyArray<E>,
  library: {
    [K in T | E]?: ReadonlyArray<Exclude<T | E, K>>;
  }
}
```

Here's a [playground link](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=8&pc=1#code/C4TwDgpgBAIhkDsAmEEGMQHEBOBDMAFgDwAqUEAHsKkgM5S3DYCWCA5lALwNOtsA0UAKLkqNemQA+PFuy4y+APnkBvAFBRN5BE3AB7VsABcUAEoRcSPQgA2IAILY8IIkMX8NWm8wBGebCAm6lohUADaANJQrFBSwgC6APwm5pbWdo7OrhRoNgCuKKRQ0kKCEYqKANyemgC+avVqAGZ56MDM1lBWAMp6ALYQwAR8RZTUyPSMsgLCouN0scUK7IoAFGwmcIgo6Fh4hKSCbgCUJgBuBkhQKo09-YPD7KvBWqi6YAY6JmEA5E16eh+gh+Plw2CBUBBuAAXj94h4Qt4-GDAtcaqF-npvj8AI55PIUOEIrT1WrHNRAA).

Reply:
> properly represent separate dependency trees of libraries.

I think the thing missing would be a mapping between library name in the tree and its identifier in the source code, like

```
  root_dependencies: ['foo', 'bar']
  library_dependencies: {
    foo: [{ library: 'bar_v0.5', name: 'bar' }]
  }
```
both root and foo depend on bar (named bar in their respective sources) but on differnt versions. Did I understand the issue correctly?

Reply:
I think that this is covered by the next incarnation of https://github.com/noir-lang/noir/pull/2997 as we can differentiate between versions based on their entrypoint.